rust-src/id/src/ffi.rs

```diff
@@ -387,6 +583,73 @@ mod test {
             &ip_cdi_secret_key,
         );
 
+        // Test ip_info_create.
+        let ip_info_copy = {
+            // Copy the IpInfo instance.
+            let ip_info_ptr = Box::into_raw(Box::new(ip_info.clone()));
+
+            // Get the identity.
+            let ip_info_identity = ip_info_ip_identity(ip_info_ptr);
+
+            // Get the description fields.
+            let mut ip_info_desc_bs_len: size_t = 0;
+            let ip_info_desc: Description = {
+                let ip_info_desc_ptr = ip_info_description(ip_info_ptr, &mut ip_info_desc_bs_len);
+                let buf = &mut slice_from_c_bytes!(ip_info_desc_ptr, ip_info_desc_bs_len as usize);
+                from_bytes(buf).expect("Bad stuff happened.")
+                // FIXME: How to convert the description byte array pointer into
+                // a Description instance? The byte array is contructed using:
+                //
+                //           description.description.as_bytes().to_vec();
+                //
+                // where description has type Description. The following did not
+                // work for me:
+                //
+                //           from_bytes::<Description, _>(buf)
+                //
+                // and fails with 'ffi::test::test_pipeline' panicked at 'Bad
+                // stuff happened: failed to fill whole buffer'.
+            };
+
+            let ip_info_name_bs = ip_info_desc.name.as_bytes();
+            let ip_info_url_bs = ip_info_desc.url.as_bytes();
+            let ip_info_desc_bs = ip_info_desc.description.as_bytes();
+
+            // Keys.
+            let mut ip_info_verify_key_len: size_t = 0;
+            let ip_info_verify_key_ptr =
+                ip_info_verify_key(ip_info_ptr, &mut ip_info_verify_key_len);
+
+            // Identity provider cdi verify key.
+            let mut ip_info_cdi_verify_key_len: size_t = 0;
+            let ip_info_cdi_verify_key_ptr =
+                ip_info_cdi_verify_key(ip_info_ptr, &mut ip_info_cdi_verify_key_len);
+
+            unsafe {
+                ip_info_create(
+                    ip_info_identity,
+                    ip_info_verify_key_ptr,
+                    ip_info_verify_key_len,
+                    ip_info_cdi_verify_key_ptr,
+                    ip_info_cdi_verify_key_len,
+                    ip_info_name_bs.as_ptr(),
+                    ip_info_name_bs.len(),
+                    ip_info_url_bs.as_ptr(),
+                    ip_info_url_bs.len(),
+                    ip_info_desc_bs.as_ptr(),
+                    ip_info_desc_bs.len(),
+                )
+            }
+        };
+
+        // FIXME: How to check equality; can I simple derive PartialEq and Eq,
+        //        and it will compare the field values (recursively) using
+        //        their comparison functions?
```

Comment:
Yes.

Reply:
Thanks. I still find it odd that I can simply compare the ps_sig::PublicKey<P> instances when P is Bls12.

But in that case I will scrap it; I wanted to ensure the FFI functions work as intended since I am not very experienced in Rust.

Reply:
We manually implement it
```
impl<C: Pairing> PartialEq for PublicKey<C> {
    fn eq(&self, other: &Self) -> bool {
        self.g == other.g
            && self.g_tilda == other.g_tilda
            && self.ys == other.ys
            && self.y_tildas == other.y_tildas
            && self.x_tilda == other.x_tilda
    }
}

```
giving the constraint `C: Pairing`. If that is what you mean.